pkg/serverless/serverless.go

```diff
@@ -263,58 +291,83 @@ func WaitForNextInvocation(stopCh chan struct{}, daemon *Daemon, metricsChan cha
 		return fmt.Errorf("WaitForNextInvocation: can't unmarshal the payload: %v", err)
 	}
 
-	if payload.EventType == "INVOKE" {
-		log.Debug("Received invocation event")
-		daemon.ComputeGlobalTags(payload.InvokedFunctionArn, config.GetConfiguredTags(true))
-		aws.SetARN(payload.InvokedFunctionArn)
-		daemon.StartInvocation()
-		if coldstart {
-			ready := daemon.WaitUntilClientReady(clientReadyTimeout)
-			if ready {
-				log.Debug("Client library registered with extension")
-			} else {
-				log.Debug("Timed out waiting for client library to register with extension.")
-			}
-			daemon.UpdateStrategy()
-		}
-
-		// immediately check if we should flush data
-		// note that since we're flushing synchronously here, there is a scenario
-		// where this could be blocking the function if the flush is slow (if the
-		// extension is not quickly going back to listen on the "wait next event"
-		// route). That's why we use a context.Context with a timeout `flushTimeout``
-		// to avoid blocking for too long.
-		// This flushTimeout is re-using the forwarder_timeout value.
-		if daemon.flushStrategy.ShouldFlush(flush.Starting, time.Now()) {
-			log.Debugf("The flush strategy %s has decided to flush the data in the moment: %s", daemon.flushStrategy, flush.Starting)
-			flushTimeout := config.Datadog.GetDuration("forwarder_timeout") * time.Second
-			ctx, cancel := context.WithTimeout(context.Background(), flushTimeout)
-			daemon.TriggerFlush(ctx, false)
-			cancel() // free the resource of the context
-		} else {
-			log.Debugf("The flush strategy %s has decided to not flush in the moment: %s", daemon.flushStrategy, flush.Starting)
-		}
-		daemon.WaitForDaemon()
+	if payload.EventType == Invoke {
+		invoke(daemon, payload.InvokedFunctionArn, payload.DeadlineMs, safetyBufferTimout, coldstart, handleInvocation)
 	}
-	if payload.EventType == "SHUTDOWN" {
+	if payload.EventType == Shutdown {
 		log.Debug("Received shutdown event. Reason: " + payload.ShutdownReason)
-
-		if strings.ToLower(payload.ShutdownReason) == "timeout" {
+		isTimeout := strings.ToLower(payload.ShutdownReason.String()) == Timeout.String()
+		if isTimeout {
 			metricTags := addColdStartTag(daemon.extraTags)
 			sendTimeoutEnhancedMetric(metricTags, metricsChan)
 		}
-
-		daemon.Stop()
+		daemon.Stop(isTimeout)
 		stopCh <- struct{}{}
 	}
 
 	return nil
 }
 
+func invoke(daemon *Daemon, arn string, deadlineMs int64, safetyBufferTimout time.Duration, coldstart bool, invocationHandler InvocationHandler) {
```

Comment:
Looking at these two functions (`invoke` and `handleInvocation`) it's not obvious from the names what exactly each one does. I would add docstring comments above each function, and also consider renaming `invoke` to `callInvocationHandler`, because the `invoke` function doesn't actually "invoke" anything per se.

Reply:
Also, could we explore adding a test for the `invoke` function? Though I am personally not sure how to go about testing a function like this...

Reply:
Never mind, looks like there is already a test for it 🙂

Reply:
✅